src/Test/L1/Worker/WorkerL1Tests.cs

```diff
@@ -0,0 +1,188 @@
+            message.Steps.Add(failStep);
+
+            // Act
+            var results = await RunWorker(message);
+
+            // Assert
+            AssertJobCompleted();
+            Assert.Equal(TaskResult.Succeeded, results.Result);
+
+            var steps = GetSteps();
+            Assert.Equal(4, steps.Count()); // Init, CmdLine, CmdLine, Finalize
+            var faiLStep = steps[2];
+            Assert.Equal(TaskResult.Skipped, faiLStep.Result);
+        }
+
+        [Fact]
+        [Trait("Level", "L1")]
+        [Trait("Category", "Worker")]
+        public async Task StepTarget_RestrictedMode()
+        {
+            // Arrange
+            var message = LoadTemplateMessage();
+            // Remove all tasks
+            message.Steps.Clear();
+            var tagStep = CreateScriptTask("echo \"##vso[build.addbuildtag]sometag\"");
+            tagStep.Target = new StepTarget
+            {
+                Commands = "restricted"
+            };
+            message.Steps.Add(tagStep);
+
+            // Act
+            var results = await RunWorker(message);
+
+            // Assert
+            AssertJobCompleted();
+            Assert.Equal(TaskResult.Succeeded, results.Result);
+
+            var steps = GetSteps();
+            Assert.Equal(3, steps.Count()); // Init, CmdLine, Finalize
+            var log = GetTimelineLogLines(steps[1]);
+            Assert.Equal(1, log.Where(x => x.Contains("##vso[build.addbuildtag] is not allowed in this step due to policy restrictions.")).Count());
+            Assert.Equal(0, GetMockedService<FakeBuildServer>().BuildTags.Count);
+        }
+
+        // Enable this test when read only variable enforcement is added
+        /*[Fact]
+        [Trait("Level", "L1")]
+        [Trait("Category", "Worker")]
+        public async Task Readonly_Variables()
+        {
+            // Arrange
+            var message = LoadTemplateMessage();
+            // Remove all tasks
+            message.Steps.Clear();
+            // Add a normal step and one that only runs on failure
+            message.Steps.Add(CreateScriptTask("echo ##vso[task.setvariable variable=system]someothervalue"));
+            var alwayStep = CreateScriptTask("echo SystemVariableValue=$(system)");
+            alwayStep.Condition = "always()";
+            message.Steps.Add(alwayStep);
+
+            // Act
+            var results = await RunWorker(message);
+
+            // Assert
+            AssertJobCompleted();
+            Assert.Equal(TaskResult.Succeeded, results.Result);
+
+            var steps = GetSteps();
+            Assert.Equal(4, steps.Count()); // Init, CmdLine, CmdLine, Finalize
+
+            var failToSetStep = steps[1];
+            Assert.Equal(TaskResult.Failed, failToSetStep.Result);
+
+            var outputStep = steps[2];
+            var log = GetTimelineLogLines(outputStep);
+            Assert.True(log.Where(x => x.Contains("SystemVariableValue=build")).Count() == 1);
+        }*/
+    }
+}
```

Comment:
this is a good start, but part of this process should be to identify the test matrix we want for L1s and then prioritize the ones that we want to get done now.

Reply:
I broke up the tests into their own classes based on categories. We don't want to bloat the PR with too many additional tests. We plan on adding new ones in separate PRs once this is merged.